Configurable Message.Key and Message.Value bytes allocation functions

colega opened the pull request:
I've seen in the profiles that a big amount of allocations come from creating the slices for `Key` and `Value` fields of `Message` through `readNewBytes`.

I can understand that we don't want to make clients accountable for properly releasing those slices by default, as that can be an unnecessary toil for many. However, in high performance applications it would be nice to have control over this.

colega commented:
I don't know why tests failed, it doesn't seem related to my PR.

rhansen2 commented:
Thanks for your patience with this PR. Could you include an example in the docs for using this new feature?

vasily-kirichenko commented:
It would be great if you added an example (or test) of using a custom key/value allocator. Would it be a `sync.Pool`-based cache or something? When to return a used array back to the pool?

colega commented on read.go:
```diff
@@ -141,7 +141,9 @@ func readBytesWith(r *bufio.Reader, sz int, cb func(*bufio.Reader, int, int) (in
 }
 
-func readNewBytes(r *bufio.Reader, sz int, n int) ([]byte, int, error) {
+// readIntoAllocatedBytes reads n bytes from r into a slice provided by makeBytes, which should should provide a slice with the required length.
```
Fixed in https://github.com/segmentio/kafka-go/pull/1104/commits/036e2327fb359dad5aecb4f94c327b765fb2dbc5

colega commented:
Hey, sorry for the delay, I've added a benchmark that shows the usage of these new fields as an example. Would that be enough? I think this is a niche feature that is only relevant in really high-volume environments, so I wouldn't like to distract the newcomers to the landing page with this.

I also fixed the benchmark which was using wrong offset to consume.

Finally the benchmark results:
```
$ benchstat -row .name -col /pooling kafka-bytes-pool
goos: darwin
goarch: arm64
pkg: github.com/segmentio/kafka-go
       │  disabled   │            enabled            │
       │   sec/op    │   sec/op     vs base          │
Reader   4.790µ ± 1%   4.739µ ± 3%  ~ (p=0.239 n=10)

       │   disabled   │            enabled             │
       │     B/s      │     B/s       vs base          │
Reader   407.8Mi ± 1%   412.1Mi ± 3%  ~ (p=0.247 n=10)

       │   disabled   │              enabled               │
       │     B/op     │    B/op     vs base                │
Reader   2089.00 ± 0%   63.50 ± 2%  -96.96% (p=0.000 n=10)

       │  disabled  │            enabled             │
       │ allocs/op  │ allocs/op   vs base            │
Reader   2.000 ± 0%   2.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```
